composer/algorithms/seq_length_warmup/seq_length_warmup.py

```diff
@@ -65,10 +78,48 @@ def set_batch_sequence_length(batch: Dict[str, torch.Tensor], curr_seq_len: int,
 
     assert isinstance(batch, Mapping)
 
+    # This should act like a no-op if curr_seq_len isn't shorter than the batch's sequence length
+    batch_seq_len = 0
+    # The batch sequence length is assumed to be the shape[1] dimension of any non-1D batch tensor
+    for batch_tensor in batch.values():
+        tensor_shape = batch_tensor.shape
+        if len(tensor_shape) > 1:
+            batch_seq_len = tensor_shape[1]
+            break
+    if curr_seq_len >= batch_seq_len:
+        return batch
+
     if truncate:
-        for k in batch.keys():
-            batch[k] = batch[k][:, :curr_seq_len]
+        # Truncate, but preserve end-of-sequence tokens
+        if preserve_end_of_sequence:
+            if 'attention_mask' not in batch:
+                raise ValueError(
+                    'Sequence Length Warmup requires that the batch has "attention_mask" when using ``preserve_end_of_sequence=True``.'
+                )
+            r_idx = torch.arange(batch['attention_mask'].shape[0])
+            # eos_idx should point to the final token index for each batch sample
+            eos_idx = batch['attention_mask'].sum(dim=1).long() - 1
+            # eos_idx_truncated is the same thing, after truncation is applied
+            eos_idx_truncated = eos_idx.clamp(max=curr_seq_len - 1)
+
+            for k in batch.keys():
+                if len(batch[k].shape) < 2:
```

Comment:
```suggestion
                if batch[k].ndim < 2:
```

 Can we use `ndim` here instead?

Reply:
Yes! Great suggestion. I will change these throughout. (I will incorporate all these changes in a single commit and resolve the individual suggestions separately.)